flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/TypeInferenceExtractor.java

```diff
@@ -143,6 +145,27 @@ public static TypeInference forAsyncTableFunction(
         return extractTypeInference(mappingExtractor);
     }
 
+    public static TypeInference forProcedure(
+            DataTypeFactory typeFactory, Class<? extends Procedure> procedure) {
+        final ProcedureMappingExtractor mappingExtractor =
+                new ProcedureMappingExtractor(
+                        typeFactory,
+                        procedure,
+                        ProcedureDefinition.PROCEDURE_CALL,
+                        ProcedureMappingExtractor.createParameterSignatureExtraction(1),
+                        ProcedureMappingExtractor.createReturnTypeResultExtraction(),
+                        ProcedureMappingExtractor.createParameterAndReturnTypeVerification());
+        try {
+            return extractTypeInferenceOrError(mappingExtractor);
+        } catch (Throwable t) {
+            throw extractionError(
+                    t,
+                    "Could not extract a valid type inference for procedure class '%s'. "
+                            + "Please check for implementation mistakes and/or provide a corresponding hint.",
+                    mappingExtractor.getFunctionClass().getName());
+        }
+    }
```

Comment:
Good suggestion.

Reply:
Good suggestion. But it may be not as so good to extract a common method considering we also want to throw the exception that it's an error in type inference for procedure or function. If a common method is extracted, we may need to use if/else or another flag to mark it which I think is not as so good.
But I would like to extract this snippet to a method to follow the style of type inference for function.